autogen/oai/together.py

```diff
@@ -0,0 +1,369 @@
+
+            together_params["stream"] = False
+
+        # To support avoidance of tool calling we allow the user to exclude tool calls on certain conditions
+        together_params["hide_tools"] = validate_parameter(
+            params, "hide_tools", str, False, "never", None, ["if_all_run", "if_any_run", "never"]
+        )
+
+        return together_params
+
+    def create(self, params: Dict) -> ChatCompletion:
+
+        messages = params.get("messages", [])
+
+        # Convert AutoGen messages to Together.AI messages
+        together_messages = oai_messages_to_together_messages(messages)
+
+        # Parse parameters to Together.AI API's parameters
+        together_params = self.parse_params(params)
+
+        # Handle tool use
+        tools = params.get("tools", None)
+        hide_tools = should_hide_tools(together_messages, tools, together_params["hide_tools"])
+
+        # We use chat model by default
+        client = Together(api_key=self.api_key)
+
+        # Token counts will be returned
+        prompt_tokens = 0
+        completion_tokens = 0
+        total_tokens = 0
+
+        max_retries = 5
+        for attempt in range(max_retries):
+            ans = None
+            try:
+                response = client.chat.completions.create(
+                    model=together_params["model"],
+                    max_tokens=together_params["max_tokens"],
+                    temperature=together_params["temperature"],
+                    top_p=together_params["top_p"],
+                    top_k=together_params["top_k"],
+                    repetition_penalty=together_params["repetition_penalty"],
+                    presence_penalty=together_params["presence_penalty"],
+                    frequency_penalty=together_params["frequency_penalty"],
+                    min_p=together_params["min_p"],
+                    stream=together_params["stream"],
+                    messages=together_messages,  # Main messages
+                    tool_choice=("auto" if not hide_tools else None),
+                    tools=tools if not hide_tools else None,  # Include any tools/functions
+                    n=1,  # API supports more than one response, however we will limit it to one
+                    safety_model=together_params["safety_model"],
+                )
+            except error.AuthenticationError as e:
+                raise RuntimeError(
+                    f"Together.AI AuthenticationError, ensure you have your Together.AI API key set: {e}"
+                )
+            except error.RateLimitError as e:
+                raise RuntimeError(f"Together.AI RateLimitError, too many requests sent in a short period of time: {e}")
+            except error.InvalidRequestError as e:
+                raise RuntimeError(f"Together.AI InvalidRequestError: {e}")
+            except Exception as e:
+                raise RuntimeError(f"Together.AI exception occurred while calling Together.API: {e}")
+            else:
```

Comment:
I don't see the point of catching Togther.AI's exception and then raise it. We should just remove this.

Reply:
Sure, I've reduced it to a single exception capture and raise.
'''
            except Exception as e:
                raise RuntimeError(f"Together.AI exception occurred: {e}")
'''